subsys/bluetooth/host/adv.c

```diff
@@ -27,6 +27,121 @@ enum adv_name_type {
 	ADV_NAME_TYPE_SD,
 };
 
+struct bt_ad {
+	const struct bt_data *data;
+	size_t len;
+};
+
+struct ad_stream {
+	/* ad is a two dimensional array of struct bt_data elements. */
+	const struct bt_ad *ad;
+	/* The number of struct bt_ad elements. */
+	size_t ad_len;
+
+	/* The current index in the array of struct bt_ad elements */
+	size_t ad_index;
+	/* The current index in the array of ad.data elements */
+	size_t data_index;
+
+	/* Current LTV offset contains the offset in the AD structure.
+	 * The length and type are included in this offset.
+	 */
+	uint16_t current_ltv_offset;
+};
+
+static void ad_stream_new(struct ad_stream *stream, const struct bt_ad *ad, size_t ad_len)
+{
+	(void)memset(stream, 0, sizeof(*stream));
+	stream->ad = ad;
+	stream->ad_len = ad_len;
+}
+
+/** @brief Returns true if the current stream is empty.
+ *
+ * @param stream AD stream, @ref ad_stream_new
+ *
+ *
+ * @returns true if the stream is now empty.
+ */
+static bool ad_stream_is_empty(const struct ad_stream *stream)
+{
+	return stream->ad_index == stream->ad_len;
+}
+
+/** @brief Returns the bt_data structure that is currently being read
+ *
+ * If the structure has been fully read, the function iterates to the next
+ *
+ * @param stream AD stream, @ref ad_stream_new
+ *
+ * @returns The current LTV structure or NULL if there are no left.
+ */
+static const struct bt_data *ad_stream_current_ltv_update(struct ad_stream *stream)
+{
+	const struct bt_data *current_ltv = &stream->ad[stream->ad_index].data[stream->data_index];
+	const bool done_reading_ltv = (stream->current_ltv_offset == current_ltv->data_len + 2);
+
+	if (done_reading_ltv) {
+		stream->current_ltv_offset = 0;
+
+		if (stream->data_index + 1 == stream->ad[stream->ad_index].len) {
+			stream->data_index = 0;
+			stream->ad_index++;
+		} else {
+			stream->data_index++;
+		}
+	}
+
+	if (ad_stream_is_empty(stream)) {
+		return NULL;
+	} else {
+		return &stream->ad[stream->ad_index].data[stream->data_index];
+	}
+}
+
+/** @brief Read at max buf_len data from the flattened AD stream.
+ *
+ * The read data can contain multiple LTV AD structures.
+ *
+ * @param stream  AD stream, @ref ad_stream_new
+ * @param buf     Buffer where the data will be put
+ * @param buf_len Buffer length
+ *
+ * @returns The number of bytes read from the stream written to the provided buffer
+ */
+static uint8_t ad_stream_read(struct ad_stream *stream, uint8_t *buf, uint8_t buf_len)
```

Comment:
Suggestion: We could actually also have used the `struct net_buf_simple` instead of supplying a `buf`, `buf_len` and returning the number of bytes read, as the `net_buf_simple` API could do that for us. 

It is OK to keep as is, but something to consider.

Reply:
I will keep it as is for now. The PR for allowing to set long periodic adv data has been open for two months now, I would like to get the functionality merged. I see that this structure can be reused in other places. I suggest to do that refactoring then